### PR TITLE
BMS Quadruna Port

### PR DESCRIPTION
**Pull request by will-chaba**
### Summary

Thruna BMS Code ported to Quadruna board. Two main peripherals have changed, those being the charger and the switch from EEPROM to SD card, as such these functionalities have been disabled until they can be ported over. Jira tickets created.

Bootloader support is also not present, but a jira ticket has been created and this will be done at a later date.

### Changelist 

- BMS Folder added under Quadruna
- BMS build targets for Thruna/Quadruna setup 
- BMS  IOC's renamed to match target car

### Testing Done

Tested both on BMS Core v3.0 and on GTest. Able to boot into state machine and send and recieve CAN messages.

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: firmware/quadruna/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# add_subdirectory(BMS)
-add_subdirectory(FSM)
+add_subdirectory(BMS)
+# add_subdirectory(FSM)
```
Comment (will-chaba) [inline firmware/quadruna/CMakeLists.txt:2]: This will be added back after #1145

File: .vscode/settings.json [hunk 2]
```diff
@@ -8,5 +8,10 @@
 		}
 	],
 	"cmake.configureOnEdit": false,
-	"cmake.configureOnOpen": false
+	"cmake.configureOnOpen": false,
+	"files.associations": {
+		"": "",
+		"hw_can.h": "c",
+		"app_imd.h": "c"
+	}
```
Comment (gtaharaedmonds) [inline .vscode/settings.json:16]: Can you revert
(I have a PR to delete this file from GitHub so this'll stop happening, sorry!)

File: .vscode/launch.json [hunk 3]
```diff
@@ -17,7 +17,37 @@
...
+            "port": 0,
+            "type": "console"
+          }
+        ]
+      },
+    },
+    {
+      "name": "Debug Embedded: BMS (Quadruna)",
+      "cwd": "${workspaceRoot}",
+      "executable": "build_fw_deploy/firmware/quadruna/BMS/BMS_quadruna.hex",
```
Comment (gtaharaedmonds) [inline .vscode/launch.json:37]: `quadruna_BMS.hex` in tasks.json but `BMS_quadruna.hex` here

File: can_bus/quadruna/BMS/BMS_alerts.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        "StackWaterMarkHighTaskCanRx",
+        "StackWaterMarkHighTaskCanTx",
+        "WatchdogTimeout",
+        "TxOverflow",
+        "RxOverflow"
+    ],
+    "faults": [
+        "MissingFSMHeartbeat",
+        "MissingDCMHeartbeat",
+        "MissingPDMHeartbeat",
```
Comment (gtaharaedmonds) [inline can_bus/quadruna/BMS/BMS_alerts.json:19]: Board names have changed

File: firmware/quadruna/BMS/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+set(HW_INCLUDE_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/src/hw" "${CMAKE_CURRENT_SOURCE_DIR}/src" "${SHARED_HW_INCLUDE_DIR}")
+
+# tests
+file(GLOB_RECURSE TEST_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/test/*.cpp")
+set(TEST_INCLUDE_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/test")
+list(APPEND TEST_INCLUDE_DIRS "${SHARED_TEST_UTILS_INCLUDE_DIRS}")
+
+# commit info
+set(COMMIT_INFO_HEADER_PATH "${CMAKE_CURRENT_BINARY_DIR}/app/app_commitInfo.h")
+set(COMMIT_INFO_SRC_PATH "${CMAKE_CURRENT_BINARY_DIR}/app/app_commitInfo.c")
```
Comment (gtaharaedmonds) [inline firmware/quadruna/BMS/CMakeLists.txt:53]: don't think you actually use these anywhere?

File: firmware/quadruna/BMS/src/cubemx/Src/stm32h7xx_it.c [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+ * @brief This function handles Hard fault interrupt.
+ */
+void HardFault_Handler(void)
+{
+    /* USER CODE BEGIN HardFault_IRQn 0 */
+
+    /* USER CODE END HardFault_IRQn 0 */
+    while (1)
+    {
+        /* USER CODE BEGIN W1_HardFault_IRQn 0 */
```
Comment (gtaharaedmonds) [inline firmware/quadruna/BMS/src/cubemx/Src/stm32h7xx_it.c:95]: Can you add hardfaulthandler here? See other boards as reference
Makes debugging hard faults a lot easier

**Review by gtaharaedmonds — COMMENTED** [review]
LFGGGG a few minor changes and then let's send it

File: can_bus/quadruna/BMS/BMS_alerts.json (hunk 4 above)
Comment (will-chaba) [reply can_bus/quadruna/BMS/BMS_alerts.json:19]: Will update this when HeartbeatBoards enum is updated, as will need to leave other sections of code as TODO until then anyways

**Review by gtaharaedmonds — APPROVED** [review]
yeep

**Pull request merged**